plugins/importtext_procedure.cc

```diff
@@ -859,54 +862,93 @@ struct ImportTextProcedureWorkInstance
+                if (!areOutputColumnNamesKnown) {
+
+                    //we need the first line to initialize
+                    ExcAssert(names != nullptr);
+                    for (unsigned i = 0;  i < numVals;  ++i) {
+                        const ColumnName & c = names[i];
+                        ColumnHash ch(c);
+                        knownColumnNames.push_back(c);
+                        if (!columnIndex.insert(make_pair(ch, i)).second)
+                            throw HttpReturnException(400, "Duplicate column name in CSV file",
+                                                      "columnName", c.toString());
+                    }
+                    dataset->initialize(knownColumnNames, columnIndex);
+                    areOutputColumnNamesKnown = true;
+                }
+            }
+
+            if (!names) {
+                threadAccum.add(actualLineNum, std::move(rowName), rowTs, vals);
+            }
+            else {
+
+                if (numVals != knownColumnNames.size())
+                    throw HttpReturnException(400, "Variable number of columns while importing text to tabular dataset");
 
-	            threadAccum.add(actualLineNum, std::move(rowName), rowTs, vals);
+                    std::vector<CellValue> orderedValues(knownColumnNames.size());
+                    for (int i = 0; i < numVals; ++i) {
 
-	            if (threadAccum.rowCount() == ROWS_PER_CHUNK) {
-	                //size_t before JML_UNUSED = threadAccum.memusage();
-	                threadAccum.freeze();
-	                //size_t after JML_UNUSED = threadAccum.memusage();
-	                TabularDatasetChunk newChunk(numberOutputColumns, ROWS_PER_CHUNK);
-	                std::unique_lock<std::mutex> guard(lineMutex);
-	                doneChunks.emplace_back(std::move(newChunk));
-	                doneChunks.back().swap(threadAccum);
-	                ExcAssertEqual(threadAccum.rowCount(), 0);
+                       auto iter = columnIndex.find(names[i]);
+                       if (iter == columnIndex.end())
+                          throw HttpReturnException(400, "Inconsistent column names while importing text to tabular dataset");
+
+                       orderedValues[iter->second] = vals[i];
+                    }
+
+                    threadAccum.add(actualLineNum, std::move(rowName), rowTs, &orderedValues[0]);
+                }
+
+	        if (threadAccum.rowCount() == ROWS_PER_CHUNK) {
+	            //size_t before JML_UNUSED = threadAccum.memusage();
+	            threadAccum.freeze();
+	            //size_t after JML_UNUSED = threadAccum.memusage();
+	            TabularDatasetChunk newChunk(numVals, ROWS_PER_CHUNK);
+	            std::unique_lock<std::mutex> guard(lineMutex);
+	            doneChunks.emplace_back(std::move(newChunk));
+	            doneChunks.back().swap(threadAccum);
+	            ExcAssertEqual(threadAccum.rowCount(), 0);
 
 	#if 0
```

Comment:
While you are _intimate_ with the code, consider removing this commented code if it no longer apply.